docs/usage/with-typescript.md

```diff
@@ -0,0 +1,224 @@
+
+## createAction
+
+For most use cases, there is no need to have a literal definition of `action.type`, so the following can be used:
+
+```typescript
+createAction<number>('test')
+```
+
+This will result in the created action being of type `PayloadActionCreator<number, string>`.
+
+In some setups, you will need a literal type for `action.type`, though.
+Unfortunately, TypeScript type definitions do not allow for a mix of manually-defined and inferred type parameters, so you'll have to specify the `type` both in the Generic definition as well as in the actual JavaScript code:
+
+```typescript
+createAction<number, 'test'>('test')
+```
+
+If you are looking for an alternate way of writing this without the duplication, you can use a prepare callback so that both type parameters can be inferred from arguments, removing the need to specify the action type.
+
+```typescript
+function withPayloadType<T>() {
+  return (t: T) => ({ payload: t })
+}
+createAction('test', withPayloadType<string>())
+```
+
+### Alternative to using a literally-typed `action.type`
+
+If you are using `action.type` as discriminator on a discriminated union, for example to correctly type your payload in `case` statements, you might be interested in this alternative:
+
+Created action creators have a `match` method that acts as a [type predicate](https://www.typescriptlang.org/docs/handbook/advanced-types.html#using-type-predicates):
+
+```typescript
+const increment = createAction<number>('increment')
+function test(action: Action) {
+  if (increment.match(action)) {
+    // action.payload inferred correctly here
+    action.payload
+  }
+}
+```
+
+This `match` method is also very useful in combination with `redux.observable` and rxjs's filter method.
+
+## createReducer
+
+The default way of calling `createReducer` would be with a map object, like this:
+
+```typescript
+createReducer(0, {
+  increment: (state, action: PayloadAction<number>) => state + action.payload
+})
+```
+
+Unfortunately, as the keys are only strings, using that API TypeScript can neither infer nor validate the action types for you:
+
+```typescript
+{
+  const increment = createAction<number, 'increment'>('increment')
+  const decrement = createAction<number, 'decrement'>('decrement')
+  createReducer(0, {
+    [increment.type]: (state, action) => {
+      // action is any here
+    },
+    [decrement.type]: (state, action: PayloadAction<string>) => {
+      // even though action should actually be PayloadAction<number>, TypeScript can't detect that and won't give a warning here.
+    }
+  })
+}
+```
+
+This is why we're offering an alternative API.
+
+### Recommendation for TS users: the alternative builder API
+
+Instead of using a Map Object as an argument to `createReducer`, you can also use a callback that receives a `ActionReducerMapBuilder` object:
```

Comment:
```suggestion
Instead of using a simple object as an argument to `createReducer`, you can also use a callback that receives a `ActionReducerMapBuilder` instance:
```